lib/libc/gen/devname.3
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)devname.3	8.2 (Berkeley) 4/29/95
.\" $FreeBSD: releng/11.1/lib/libc/gen/devname.3 188497 2009-02-11 20:24:59Z ed $
.\"
.Dd February 22, 2005
.Dt DEVNAME 3
.Os
.Sh NAME
.Nm devname
.Nd "get device name"
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/stat.h
.In stdlib.h
.Ft char *
.Fn devname "dev_t dev" "mode_t type"
.Ft char *
.Fn devname_r "dev_t dev" "mode_t type" "char *buf" "int len"
.Ft char *
.Fn fdevname "int fd"
.Ft char *
.Fn fdevname_r "int fd" "char *buf" "int len"
.Sh DESCRIPTION
The
.Fn devname
function returns a pointer to the name of the block or character
device in
.Pa /dev
with a device number of
.Fa dev ,
and a file type matching the one encoded in
.Fa type
which must be one of
.Dv S_IFBLK
or
.Dv S_IFCHR .
To find the right name,
.Fn devname
asks the kernel via the
.Va kern.devname
sysctl.
If it is unable to come up with a suitable name,
it will format the information encapsulated in
.Fa dev
and
.Fa type
in a human-readable format.
.Pp
The
.Fn fdevname
and
.Fn fdevname_r
function obtains the device name directly from a file descriptor
pointing to a character device.
If it is unable to come up with a suitable name, these functions will
return a NULL pointer.
.Pp
.Fn devname
and
.Fn fdevname
return the name stored in a static buffer which will be overwritten
on subsequent calls.
.Fn devname_r
and
.Fn fdevname_r
take a buffer and length as argument to avoid this problem.
.Sh EXAMPLES
.Bd -literal -compact
int fd;
struct stat buf;
char *name;

	fd = open("/dev/tun");
	fstat(fd, &buf);
	printf("devname is /dev/%s\en", devname(buf.st_rdev, S_IFCHR));
	printf("fdevname is /dev/%s\en", fdevname(fd));
.Ed
.Sh SEE ALSO
.Xr stat 2
.Sh HISTORY
The
.Fn devname
function appeared in
.Bx 4.4 .
The
.Fn fdevname
function appeared in
.Fx 8.0 .


ChangeLog
2007-01-25 Christopher Olsen <cwolsen@domainatlantic.com>
	Cleaning up the ne2k driver for future network support. 
	 
	Previous version of UbixOS had LwIP functioning, I hope this will be true in this case. 



lib/libc/gen/devname.c
/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)devname.c	8.2 (Berkeley) 4/29/95";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/11.1/lib/libc/gen/devname.c 298226 2016-04-18 21:05:15Z avos $");

#include <sys/param.h>
#include <sys/sysctl.h>

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

char *
devname_r(dev_t dev, mode_t type, char *buf, int len)
{
	int i;
	size_t j;

	if (dev == NODEV || !(S_ISCHR(type) || S_ISBLK(dev))) {
		strlcpy(buf, "#NODEV", len);
		return (buf);
	}

	if (S_ISCHR(type)) {
		j = len;
		i = sysctlbyname("kern.devname", buf, &j, &dev, sizeof (dev));
		if (i == 0)
			return (buf);
	}

	/* Finally just format it */
	snprintf(buf, len, "#%c:%#jx",
	    S_ISCHR(type) ? 'C' : 'B', (uintmax_t)dev);
	return (buf);
}

char *
devname(dev_t dev, mode_t type)
{
	static char buf[SPECNAMELEN + 1];

	return (devname_r(dev, type, buf, sizeof(buf)));
}
